Implementation of history window

mghe01 opened the pull request:
[Implementation of history window](https://github.com/mghe01/PicView/commit/fb9f7a6478eb602a1942db273ce0b8512abe39bc)

Functionality changes for you to consider.

**History Window**
Introduces a new History Window with full snapshot management, timeline navigation, and UI integration.
Actions such as Crop, Rotate, Flip, Effect & Resize are added to the history. Navigating away from an image with changes will prompt user if they want to save changes.
Added history window toggle to the Image Menu on the BottomBar and added shortcut 'H'

**Functionality Changes**
- Crop control modified to apply changes directly to the image instead of doing a save-as. Added floating menu with options to apply crop, save-as, copy to clipboard.
- Rotation and Flip modified to apply directly to image. Extended the Flip to have both horizontal and vertical flip.

**UI enhancements**
- Menu buttons enabled/disabled based on function requirements
- Navigation button disabling when end limits reached for when looping is disabled
- Startup window buttons modified so you can click on the entire button not just the text.
- Bottom bar menu icons updated to match hover menu
- Added Message Dialog and Save Changes Dialog
- Modified BottomBar context menu event handler and added context menu for Rotation button
- Added Tooltips to Bottombar
- Fixed TitleManager where it was showing rotation instead of zoom

**Export To PDF**
Added function to export the image to PDF


Other bug fixes and enhancements

mghe01 committed:
Update EditableTitlebar.axaml.cs

Right click Titlebar with no image loaded throws NullReferenceException

mghe01 committed:
Merge branch 'dev' into mghe-dev

mghe01 committed:
Merge branch 'dev' into mghe-dev

mghe01 committed:
Added Print Preview Feature

In multi-user Remote Desktop environments, the Windows Desktop Experience feature is unavailable preventing access to the built-in "Print Pictures" dialog when using the 'print' process verb. As a result, printing defaults directly to the user's default printer without allowing printer selection.

Added Enhanced printing functionality by adding print preview with support for selecting printers and print properties.

- Added Print Preview Window
- Added PrintEngine core class for unified print and preview layout handling
- Implemented shared ComputeLayout logic for consistent scaling and margins
- Integrated async RunPrintAsync with UI feedback
- Added grayscale conversion and Avalonia to GDI bitmap rendering
- Ensured 1:1 accuracy between on-screen preview and physical/PDF output

mghe01 committed:
Added Global Settings Config File

Added support for a GlobalSettings.json file in ApplicationDirectory/Config.
This file can define any settings from UserSettings.json and enforces them as global overrides at startup, regardless of values in the local or user profile UserSettings.json.

mghe01 committed:
Added Show SetAsWallpaper Setting Option

Added option in the UserSettings.json to be able to toggle the visibility of the SetAsWallpaper Menu buttons, preventing users being able to change the desktop wallpaper.

mghe01 committed:
Merge branch 'dev' into mghe-dev

mghe01 committed:
Fixed Window Maximized Overflow

When using AcrylicBlur and the Window was maximized with Auto-fit Window turned off, Avalonia continued extending into that region instead of clipping to the screen’s working area, causing the content to overflow past the screen edges.

Setting ExtendClientAreaToDecorationsHint and SystemDecorations only while maximized allowed it to reapply its normal clipping and square-corner mask, fixing the overflow.

- Set window properties SystemDecorations and ExtenClientAreatoDecorationsHint when the window is maximized
- Removed SetMargin workaround

Aditional Changes
- Adjusted bottom bar drag region to avoid overlapping the settings button
- Adjusted titlebar layout so the window title is centered horizontally

Ruben2776 committed:
Fix incorrect image disposal at startup and fix printing shortcut not working properly.

Refine file handling in `FunctionsMapper` and improve eviction logic in `Preloader`

- Fixed `null` handling for `CurrentValue?.FullName` in multiple `FunctionsMapper` methods (`Print`, `OpenWith`, `OpenInExplorer`, etc.).
- Updated `Preloader` to conditionally invoke `ImageDisposalHelper` only on successful addition (`TryAdd` eviction handling).
- Removed unnecessary validation in `FileManager.Print`.
- Corrected eviction logic in `EvictingDictionary` (`_dictionary.Count > _maxSize`).

Ruben2776 committed:
Alphabetize languages

Ruben2776 committed:
Link `PrintPreviewView.axaml.cs` to `PrintPreviewView.axaml` in project file.

Ruben2776 committed:
Update Danish translation: Change "Skala" to "Skalering" in `da.json`.

Ruben2776 committed:
Fix titlebar not properly resizing at lower widths.

mghe01 committed:
Merge branch 'dev' into mghe-dev

mghe01 committed:
Merge branch 'dev' of https://github.com/mghe01/PicView into mghe-dev

mghe01 committed:
Merge branch 'dev' into mghe-dev

mghe01 committed:
Added Rotation Context Menu in BottomBar

When the rotation button moved to the BottomBar from the TitleBar the context menu no longer work. Added it back in and tweaked the default User Control's context menu and window resize capability.

mghe01 committed:
Revert to original en.json

mghe01 committed:
Revert "Revert to original en.json"

This reverts commit 24bfe191772733d7decdb2126e0748442c5260c8.

mghe01 committed:
Reverted to original en.json

Broke the file somehow, reverted back to upstream version

mghe01 committed:
Menu Behaviour and Options

Refined menu behaviour and added multi-user configuration options

- Menu buttons now enable/disable only when an image is selected
- Added Effects window under the Topbar menu
- Introduced options: ShowPrintInBottomToolbar and EnableCheckForUpdates to improve multi-user environment support

markh-wettenhalls committed:
Merge branch 'dev' into mghe-dev

markh-wettenhalls committed:
Modify Layouts, Enhance Crop Tool, Add History Window

- Modified the StartUpMenu, and BottomBar to be responsive using xaml instead of code behind.

- Changed Crop tool to have visible menu and cropped image is shown in editor instead of calling SaveAs

- Started implementing History Window that will allow tracking changes and reverting.

markh-wettenhalls committed:
Merge branch 'dev' into mghe-dev

mghe01 committed:
Improved History Window

mghe01 committed:
History and transformation changes

mghe01 committed:
Merge branch 'dev' into mghe-dev

mghe01 committed:
merge

markh-wettenhalls committed:
Merge branch 'dev' into mghe-dev

markh-wettenhalls committed:
Fixed resizing when rotate, flipped or cropped

markh-wettenhalls committed:
Layout fixes

markh-wettenhalls committed:
Merge branch 'dev' into mghe-dev

mghe01 committed:
Merge branch 'dev' into mghe-dev

mghe01 committed:
Merge remote-tracking branch 'upstream/dev' into dev

mghe01 committed:
Rebase

mghe01 committed:
Implementation of history window

History Window
Introduces a new History Window with full snapshot management, timeline navigation, and UI integration.
Actions such as Crop, Rotate, Flip, Effect & Resize are added to the history. Navigating away from an image with changes will prompt user if they want to save changes.
Added history window toggle to the Image Menu on the BottomBar and added shortcut 'H'

Functionality Changes
- Crop control modified to apply changes directly to the image instead of doing a save-as. Added floating menu with options to apply crop, save-as, copy to clipboard.
- Rotation and Flip modified to apply directly to image. Extended the Flip to have both horizontal and vertical flip.

UI enhancements
- Menu buttons enabled/disabled based on function requirements
- Navigation button disabling when end limits reached for when looping is disabled
- Startup window buttons modified so you can click on the entire button not just the text.
- Bottom bar menu icons updated to match hover menu
- Added Message Dialog and Save Changes Dialog
- Modified BottomBar context menu event handler and added context menu for Rotation button
- Added Tooltips to Bottombar
- Fixed TitleManager where it was showing rotation instead of zoom

Export To PDF
Added function to export the image to PDF

Other bug fixes and enhancements

mghe01 committed:
Merge remote-tracking branch 'upstream/dev' into dev

mghe01 committed:
Changes to make Codacy happy with last commit

Changes to make Codacy happy with last commit

Ruben2776 committed:
Fix building for macOS

Ruben2776 commented:
First of all, thank you so much for the pull request! It looks like you put a lot of effort into it. I'm very happy to see someone willing to put so much effort into helping my hobby project for free :) 
You can ignore Codacy btw.

As this is a really large pull request, I am reluctant to immediately accepting it. As I am currently away from home, I don't have access to my Windows desktop PC, so I am unable to properly test it for the time being. 

These are some issues that I found while testing it on my MacBook:

- I made a small commit to fix it on macOS though, as there was a missing closing tag and the `_rotationContextMenu.UpdateSubscription()` threw an exception. 
- The flip animation stopped playing. Would you mind fixing it again?
- You can zoom in while the History view is open, making you automatically zoom the image and scroll the History Window at the same time.
- You replaced the old image icon, in the `Bottom Toolbar`, with a new one, while I prefer the old one to be used. Please bring it back to how it was.
- You made a popup asking if the changes should be saved every time you navigate away. As someone who hates pop-ups, this can be very infuriating. There should be an option to chose if you want it to show up.
- You removed the `Up` and `Down` keybindings. Please add the back. As it is now, you can now no longer navigate up or down, when viewing the gallery. 
- With the new Print preview window, that you previously added, I have made a dummy printer you can use to save the file as a PDF. So as it is, I prefer the export to PDF option to not be there, as I am already trying to make the UI less cluttered. 
- You have added an `await` in the `ImageLoader` inside the navigation, which means that it has to wait for your history changes to complete? Can this possibly be done after the image has been fully shown, in the background in a separate thread?
- You have a setting to disable the functionality of the `Update` - even though there is no auto update added. May I ask why?
- You changed the full screen button to be a maximize button, in the titlebar for Windows. Please bring it back to the full screen functionality or add the ability to toggle between them, in the Settings window, but having full screen as the default.

mghe01 commented:
I've started looking into the issues raised but will finalise when I get some free time. I don't have a MacOS device so thanks for fixing those issues up. I haven't been able to debug or test any changes on that platform so i try not to make any changes there.

I use PicView on our remote desktop servers which don't have a native image viewer. It's a great, easy to use option for our users and it's excellent that i can customize it. The changes I make are to improve the app and processes for our users which don't necessarily align with your vision for the product so there is no expectation for you to include any of my changes into your build. I just like to pass them on. I'll add options to toggle my changes within the settings window and any niche changes I'll just toggle from config file. 

You asked about why I added an option to disabled the updates, this was to prevent standard users from clicking it so that updates are only done manually by sys admins. Even though they don't have access rights to install the updates it's best to just disable it. The setting to disable it is only in the config file so users won't be able to toggle it.

markh-wettenhalls committed:
Added Setting options and UI Tweaks

Added Setting options for
- Maximize Window Mode
- Show Prompt To Save Changes
- Show Export To PDF In Menu

UI Tweaks

*Still looking into the other issues raised

markh-wettenhalls committed:
Added Setting options and UI Tweaks

Added Setting options for
- Maximize Window Mode
- Show Prompt To Save Changes
- Show Export To PDF In Menu

UI Tweaks

*Still looking into the other issues raised

mghe01 committed:
Merge branch 'dev' of https://github.com/mghe01/PicView into dev

Ruben2776 committed:
Update Avalonia to 11.3.10

Ruben2776 commented on src/PicView.Avalonia/ImageTransformations/Rotation/RotationTransformer.cs:
```diff
@@ -65,115 +97,55 @@ private void SetImageLayoutTransform(RotateTransform rotateTransform)
+            }
         }
 
-        switch (orientation)
```
This switch was responsible for handling EXIF rotation, I.E, JPEGs who has the correct rotation value stored within its metadata. Your current `RotationTransformer` does not rotate pictures according to EXIF values. Please download these images from https://github.com/recurser/exif-orientation-examples and check them with your version and my version to see what I am talking about.

mghe01 committed:
Navigation Fixed

Up and Down navigation in Gallery fixed

mghe01 committed:
Added Setting EnableFileChangeHistory

Added setting to toggle File Change History. When off the History Manager doesn't take any snapshots for Crop, Rotate, Flip and Effects.

mghe01 committed:
Image Transitions

mghe01 committed:
Merge remote-tracking branch 'upstream/dev' into dev

mghe01 committed:
Merge remote-tracking branch 'upstream/dev' into dev

mghe01 committed:
Flip Transition

mghe01 committed:
Merge remote-tracking branch 'upstream/dev' into dev

mghe01 committed:
Merge remote-tracking branch 'upstream/dev' into dev

mghe01 committed:
Update .gitignore

mghe01 committed:
Update EffectsView.axaml

mghe01 committed:
Merge remote-tracking branch 'upstream/dev' into dev

mghe01 committed:
Refactor and expand the Effects view UI.

Added many new sliders, toggles and numeric inputs (hue, saturation, temperature, vibrance, color-balance tabs, exposure, gamma, highlights/shadows/blacks/whites, dehaze, clarity, grain, sharpen, vignette, blur, posterize, solarize, pencil sketch, presets UI, etc.).
Add gradient brushes and UI polish (spacing, sizes, brush resources). Update code-behind and core effect classes (ImageEffectConfig, ImageEffectsHelper, EffectsView.axaml.cs) to support the new controls, parameters and preset handling.

Fixes Issue #296

Ruben2776 committed:
Merge into dev

Ruben2776 commented:
@mghe01 Thank you very much for your hard work! I'm sorry if my previous behavior was a bit rude, and I've been a bit slow to respond/review this PR. 

I have been dedicating my time and energy to rewriting the project to allow for multi-window scenarios as well tab based layouts, in the [NavigationRewrite branch](https://github.com/Ruben2776/PicView/tree/NavigationRewrite). Everything is a work in progress, including trying to make a better UI. The code is still a mess there, but once it is ready to be merged, and cleaned up, the code will be much more maintainable and sane. The `NavigationRewrite` branch is meant for a 5.0 release in the future.

I have been keeping up with this branch meanwhile though, and I am ready to post my thoughts below:

### Addressing the UI enhancements bullet points:

> Menu buttons enabled/disabled based on function requirements

You should be able to open the Image Info window regardless of the presence of a file. It is designed to automatically update when switching between files. It should also be able to show some information on pictures, that are not files, such as pasted screenshots, etc. Which I might expand upon in the future. 

The Effects window should also be able to be opened - even if there is no file viewed. Same reasoning as above, and that you should be able to apply effects to images that are not files, which could then possibly be saved with an effect applied.

> Navigation button disabling when end limits reached for when looping is disabled

That is a great idea and I have added it in my `NavigationRewrite` branch.

**Regarding the arrow enhancements:**
They look great, and I can manually add them to the dev branch, or you can submit a new PR?

_I will try to address the rest of the bullet points later._

### Regarding settings:
When I run your branch in debug mode, my settings are not loaded. If opening the Settings window, it throws a null excpetion at line 30 in SettingsManager, where it couldn't retrieve the correct  `CurrentSettingsPath` as `SettingsConfiguration? Configuration` was null.

### History window:
It seems you have given up on this feature, as it was hidden by later updates?

> **Export To PDF**

I can apply that to the `File` header, inside the top menu navigation.

### Regarding crop:
Your idea to show the extra hidden functionality is a great idea. However, I was planning to expand upon the crop view in the future, to have a small floating window show up when cropping. Where it would give you those options, as well as planning to have a dropdown menu with options such as forcing specific aspect ratios. 
I will address the image cropping changes after the `NavigationRewrite` branch has been merged.


### The new Effects window:
It looks amazing! I'd be more than happy to have it in a 4.2 release! :) Can I cherry-pick it into the dev branch - or can you make a new PR?
